sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -847,7 +847,117 @@ protected BigtableSource withEstimatedSizeBytes(Long estimatedSizeBytes) {
+          size += source.getEstimatedSizeBytes(options);
+        } else {
+          if (size > 0) {
+            reducedSplits.add(
+                new BigtableSource(
+                    config,
+                    filter,
+                    ImmutableList.of(mergeRanges(mergedRanges.build())),
+                    size));
+          }
+          counter = 1;
+          size = source.getEstimatedSizeBytes(options);
+          previousSourceRanges = source.getRanges();
+          mergedRanges = ImmutableList.builder();
+          mergedRanges.addAll(previousSourceRanges);
+        }
+        if (counter == numberToCombine) {
+          if (size > 0) {
+            reducedSplits.add(
+                new BigtableSource(
+                    config,
+                    filter,
+                    ImmutableList.of(mergeRanges(mergedRanges.build())),
+                    size));
+          }
+          counter = 0;
+          size = 0;
+          previousSourceRanges = null;
+          mergedRanges = ImmutableList.builder();
+        }
+      }
+      if (size > 0) {
+        reducedSplits.add(
+            new BigtableSource(
+                config,
+                filter,
+                ImmutableList.of(mergeRanges(mergedRanges.build())),
+                size));
+      }
+      return reducedSplits.build();
+    }
+
+    /** Helper to validate range Adjacency.
+     * Ranges are considered adjacent if [1..100][100..200][200..300]
+     **/
+    private boolean checkRangeAdjacency(List<ByteKeyRange> ranges, List<ByteKeyRange> otherRanges) {
+      checkArgument(ranges != null || otherRanges != null, "Both ranges cannot be null.");
+      ImmutableList.Builder<ByteKeyRange> mergedRanges = ImmutableList.builder();
+      if (ranges != null) {
+        mergedRanges.addAll(ranges);
+      }
+      if (otherRanges != null) {
+        mergedRanges.addAll(otherRanges);
+      }
+      return checkRangeAdjacency(mergedRanges.build());
+    }
+
+    /** Helper to validate range Adjacency.
+     * Ranges are considered adjacent if [1..100][100..200][200..300]
+     **/
+    private boolean checkRangeAdjacency(List<ByteKeyRange> ranges) {
+      int index = 0;
+      if (ranges.size() < 2) {
+        return true;
+      }
+      ByteKey lastEndKey = ranges.get(index++).getEndKey();
+      while (index < ranges.size()) {
+        ByteKeyRange currentKeyRange = ranges.get(index++);
+        if (!lastEndKey.equals(currentKeyRange.getStartKey())) {
+          return false;
+        }
+        lastEndKey = currentKeyRange.getEndKey();
+      }
+      return true;
+    }
+
+    /** Helper to combine/merge ByteKeyRange
+     * Ranges should only be merged if they are adjacent
+     * ex. [1..100][100..200][200..300] will result in [1..300]
+     * Note: this method will not check for adjacency see {@link #checkRangeAdjacency(List)}
+     **/
+    private ByteKeyRange mergeRanges(List<ByteKeyRange> ranges) {
```

Comment:
Can you please return an `ImmutableList<ByteKeyRange>` here?  That's what `reduceSplits()` needs anyway; right now there's code duplication to create the ImmutableList..

Reply:
Does static make sense here?

Reply:
Agreed will return list and since no instance variable is accessed will make method static.